sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,748 @@
+template <typename T0, typename T1, int SZ> T0 cm_reduced_max(simd<T1, SZ> v);
+template <typename T0, typename T1, int SZ> T0 cm_reduced_min(simd<T1, SZ> v);
+```
+
+### Memory access APIs
+
+Currently the variety of memory objects supported by the Explicit SIMD extension
+implementation is limited to the following:
+- USM pointers (aka 'flat address')
+- 1D global accessors
+- 2D image accessors
+
+The implementation further limits the set of memory access operations
+which a kernel can perform through those memory objects. Basically, all of them
+are special intrinsic APIs described below rather than standard SYCL accessor or
+pointer operations. Examples of unsupported features include
+`accessor::get_pointer()`, accessor's `operator []`, C/C++ dereference of an
+USM pointer, local accessors, 2D and 3D accessors.
+Those are temporary restrictions, to be dropped in future.
+
+See auto-generated documentation for the complete list of APIs here. (TBD)
+
+#### USM pointer-based memory access
+##### Flat-address gather/scatter
+perform scattered read/write for the given starting pointer **p** and
+**offsets**.
+
+```cpp
+template <typename T, int n, int ElemsPerAddr = 1,
+	  CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+typename std::enable_if<((n == 16 || n == 32) &&
+    (ElemsPerAddr == 1 || ElemsPerAddr == 2 || ElemsPerAddr == 4)),
+    simd<T, n*ElemsPerAddr>>::type
+flat_load(T *p, simd<uint32_t, n> offsets, simd<uint16_t, n> pred = 1);
+
+template <typename T, int n, int ElemsPerAddr = 1,
+	  CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+typename std::enable_if<((n == 16 || n == 32) &&
+    (ElemsPerAddr == 1 || ElemsPerAddr == 2 || ElemsPerAddr == 4)),
+    void>::type
+flat_store(T *p, simd<T, n*ElemsPerAddr> vals, simd<uint32_t, n> offsets,
+           simd<uint16_t, n> pred = 1);
+```
+##### Flat-address block load/store
+read or write a consecutive block of data for the memory location specified by
+**addr**.
+
+```cpp
+template <typename T, int n, CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+simd<T, n> flat_block_load(const T *const addr);
+
+template <typename T, int n, CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+void flat_block_store(T *p, simd<T, n> vals);
+```
+
+##### Flat-address load4/store4
+perform scattered read/write for the given starting pointer **p** and
+**offsets**. Up to 4 data elements may be accessed at each address depending on
+the enabled channel **Mask**.
+
+```cpp
+template <typename T, int n, ChannelMaskType Mask,
+          CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+    typename std::enable_if<(n == 16 || n == 32),
+                            simd<T, n * __NumChannels(Mask)>>::type
+    flat_load4(T *p, simd<uint32_t, n> offsets, simd<uint16_t, n> pred = 1);
```

Comment:
Can this be parameterized? Why is 4 magic?

Reply:
4 maps to Gen hardware memory access instruction (message)

Reply:
I understand, but is that what it is going to be like for all future hardware? Would it be wise to parameterize this so that we don't need new function names for flat_loadX at some point?

Reply:
I'm not sure. For there is 1 and 4 only, and people would have to use more verbose templated syntax anticipating other widths which is not clear if/when will happen.

Reply:
Can't this be deduced?

Reply:
Probably. Do you have an example?